### PR TITLE
Port BertEncoder from Hugging Face transformers

### PR DESCRIPTION
**Pull request by danieldk**

### CODE REVIEW

**Comment by danieldk** [comment]
The GitHub Actions failure is a fluke, it seems to come from a prior push.

File: src/bert_model.rs [hunk 1]
```diff
@@ -192,6 +193,54 @@ impl LoadFromHDF5 for BertEmbeddings {
...
+    ///
+    /// Returns the hidden states and attention per layer.
+    pub fn forward_t(&self, hidden_states: &Tensor, train: bool) -> Vec<(Tensor, Tensor)> {
+        let mut all_hidden_states = Vec::with_capacity(self.layers.len());
+
+        let mut hidden_states = CowTensor::Borrowed(hidden_states);
+        for layer in &self.layers {
+            let states_attention = layer.forward_t(&hidden_states, train);
+
+            hidden_states = CowTensor::Owned(states_attention.0.shallow_clone());
```
Comment (twuebi) [inline src/bert_model.rs:205]: Not this PR, should have been in the previous review - maybe the return type of BertLayer should have named fields?

File: src/bert_model.rs [hunk 2]
```diff
@@ -192,6 +193,54 @@ impl LoadFromHDF5 for BertEmbeddings {
...
+#[derive(Debug)]
+pub struct BertEncoder {
+    layers: Vec<BertLayer>,
+}
+
+impl BertEncoder {
+    /// Apply the encoder.
+    ///
+    /// Returns the hidden states and attention per layer.
+    pub fn forward_t(&self, hidden_states: &Tensor, train: bool) -> Vec<(Tensor, Tensor)> {
```
Comment (twuebi) [inline src/bert_model.rs:198]: See below, maybe return a `Vec<BertLayerOutput>`?

Comment (danieldk) [reply src/bert_model.rs:198]: Good idea! I'll do that in the next PR.

**Review by twuebi — APPROVED** [review]

**Pull request merged**